pytorch_lightning/core/lightning.py

```diff
@@ -91,11 +91,53 @@ def forward(self, x):
         if self.trainer.proc_rank == 0:
             print(*args, **kwargs)
 
+    # Note this is almost identical to distrib_parts.TrainerDPMixin.__transfer_data_to_device
+    # Only works for GPU and not TPU for now
+    def __transfer_data_to_device(self, batch, device):
+        # base case: nothing to do
+        if torch.is_tensor(batch) and batch.device == device:
+            return batch
+
+        # object can be directly moved using `cuda` or `to`
+        if callable(getattr(batch, 'cuda', None)) and device.type == 'cuda':
+            return batch.cuda(device=device)
+
+        if callable(getattr(batch, 'to', None)):
+            return batch.to(device=device)
+
+        # when list or tuple
+        if isinstance(batch, (list, tuple)):
+            if isinstance(batch, tuple):
+                batch = list(batch)
+            for i, x in enumerate(batch):
+                batch[i] = self.__transfer_data_to_device(x, device)
+            return batch
+
+        # when dict
+        if isinstance(batch, dict):
+            for k, v in batch.items():
+                batch[k] = self.__transfer_data_to_device(v, device)
+
+            return batch
+
+        # nothing matches, return the value as is without transform
+        return batch
+
+    def __call__(self, *input_data, **kwargs):
+        devices = [p.device for p in self.parameters()]
+        assert set(devices) == 1, 'All parameters must be on same device'
```

Comment:
rather if they are not the same type, do no transfer and just warn user

Reply:
Already changed this